keps/74-support-argo-workflow/README.md

```diff
@@ -0,0 +1,722 @@
+      parameters:
+      - name: message
+    container:
+      image: busybox
+      command: [sleep]
+      args: ["3m"]
+```
+
+After they call the CLI, the new workflow will be like:
+``` yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: loops-
+  namespace: argo
+spec:
+  entrypoint: loop-example-depth-1
+  templates:
+  - name: kueue-suspend
+    suspend: {}
+  - name: loop-example-depth-2
+    steps:
+    - - name: suspend
+        template: kueue-suspend
+    - - name: print-message-loop
+        template: print-message
+        arguments:
+          parameters:
+          - name: message
+            value: "{{item}}"
+        withItems:              # invoke print-message once for each item in parallel
+        - hello world           # item 1
+        - goodbye world         # item 2
+  - name: loop-example-depth-1
+    steps:
+    - - name: loop-example-depth-2
+        template: loop-example-depth-2
+        withItems:
+        - depth-1-1
+        - depth-1-2
+  - name: print-message
+    inputs:
+      parameters:
+      - name: message
+    container:
+      image: busybox
+      command: [sleep]
+      args: ["3m"]
+```
+
+Option 2. Integrated Suspend Capability: We propose introducing a new specification field within workflows, such 
+as suspendBySteps. If `workflow.spec.suspendBySteps` is set to true, the workflow-controller automatically 
+injects a special suspend template into each stepGroup. The kueue's job-controller monitors these and 
+generates workloads for the following steps. Upon workload admission, the suspension step is marked as 
+completed. This work can not complete with the help of workflow managers.
+
+![](suspend-with-workflow-support.drawio.svg)
+
+For example, users can submit a argo workflow like this to queue the workflow:
+``` yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: loops-
+  namespace: argo
+spec:
+  # support suspend before any step starting
+  suspendBySteps: true
+  entrypoint: loop-example-depth-1
+  templates:
+  - name: loop-example-depth-2
+    steps:
+    - - name: print-message-loop
+        template: print-message
+        arguments:
+          parameters:
+          - name: message
+            value: "{{item}}"
+        withItems:              # invoke print-message once for each item in parallel
+        - hello world           # item 1
+        - goodbye world         # item 2
+  - name: loop-example-depth-1
+    steps:
+    - - name: loop-example-depth-2
+        template: loop-example-depth-2
+        withItems:
+        - depth-1-1
+        - depth-1-2
+      - name: loop-example-depth-2-2
+        template: loop-example-depth-2
+  - name: print-message
+    inputs:
+      parameters:
+      - name: message
+    container:
+      image: busybox
+      command: [sleep]
+      args: ["3m"]
+```
+
+Option 3. Kueue Webhook Enhancement: A new webhook is added within Kueue to intercept pod creations in the cluster. 
```

Comment:
How the Argo Workflow controller will reconcile steps if Pods will be in SchedulingGated status ?

Reply:
AFAIK, SchedulingGated status is almost same with pending status, so it should not have any effect to Argo Workflow controller. Maybe @terrytangyuan can give us some more information.